platforms/android/library-compose/src/main/java/io/element/android/wysiwyg/compose/RichTextEditor.kt

```diff
@@ -0,0 +1,95 @@
+package io.element.android.wysiwyg.compose
+
+import android.text.Editable
+import androidx.compose.foundation.layout.fillMaxWidth
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCoroutineScope
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.viewinterop.AndroidView
+import androidx.core.widget.addTextChangedListener
+import io.element.android.wysiwyg.EditorEditText
+import io.element.android.wysiwyg.view.models.InlineFormat
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.emptyFlow
+import kotlinx.coroutines.launch
+import uniffi.wysiwyg_composer.ActionState
+import uniffi.wysiwyg_composer.ComposerAction
+import uniffi.wysiwyg_composer.MenuAction
+
+/**
+ * A composable rich text editor.
+ *
+ * This composable is a wrapper around the [EditorEditText] view.
+ *
+ * @param modifier The modifier for the layout
+ * @param actions Used to send actions to the editor
+ * @param onHtmlChanged Called whenever the editor content changes with the new HTML
+ * @param onMarkdownChanged Called whenever the editor content changes with the new Markdown
+ * @param onActionsChanged Called whenever the editor action states change
+ * @param onSelectionChanged Called whenever the editor selection changes
+ * @param onMenuActionChanged Called whenever the editor menu action changes
+ */
+@Composable
+fun RichTextEditor(
+    modifier: Modifier = Modifier,
+    actions: Flow<EditorAction> = emptyFlow(),
```

Comment:
This looks odd.
This kind of internal state of the RTE should be modeled with the "state holder" pattern : https://developer.android.com/jetpack/compose/state#managing-state

Reply:
Yeah it doesn't seem to be idiomatic Compose - the reason I added this flow is to work with the underlying non-Compose View. But I'd definitely prefer not to force this pattern on users of the compose library.

I thought about exposing some actions by following the pattern of `value` +`onValueChanged` i.e.
```kotlin
@Composable
fun MyComposable(
    value: String,
    onValueChange: (String) -> Unit,
)
```

But I was having difficulty propogating this down to the underlying View class without triggering all the different setters every time the update block is run.

I was also unsure about how best to extract one-shot actions such as `undo()`/`redo()` into a state.

Perhaps you know a way I can get around these issues?

Reply:
See code snippets:

https://developer.android.com/topic/architecture/ui-layer/stateholders#ui-logic

AND

https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:compose/material3/material3/src/commonMain/kotlin/androidx/compose/material3/SheetDefaults.kt;l=97?q=SheetState&sq=&ss=androidx

These show how to implement a state holder class.

Ideally that in your case that class will hold state regarding whether or not "bold" is enabled, and also offers methods to change those settings.

Then there will be a `rememberRichTextEditor()` or similar function that remembers an instance of that class. It will be used as a default value for the `RichTextEditor` composable so that, if needed, can be hoisted out of it and used by other code.

Reply:
If you can't your get around it let's book a meeting and pair a bit on it.